module/Company/src/Company/Controller/AdminController.php

```diff
@@ -0,0 +1,345 @@
+<?php
+
+namespace Company\Controller;
+
+use Zend\Mvc\Controller\AbstractActionController;
+use Zend\View\Model\ViewModel;
+
+class AdminController extends AbstractActionController
+{
+
+    public function indexAction()
+    {
+        $companyService = $this->getCompanyService();
+        
+        $vm = new ViewModel(array(
+            'companyList' => $companyService->getHiddenCompanyList()
+        ));
+        
+        return $vm;
+
+    }
+    
+    public function addPacketAction()
+    {
+        $companyService = $this->getCompanyService();
+        $packetForm = $companyService->getPacketForm();
+        $request = $this->getRequest();
+        $companyName = $this->params('slugCompanyName');    
+        if ($request->isPost()) {
+            $companyService = $this->getCompanyService();
+            $packetForm->setData($request->getPost());
+
+            // TODO: isValid does not work yet
+            //if ($companyForm->isValid()) {
+            //    $company = $companyService->insertCompany();
+            //    $company->exchangeArray($request->getPost()); // Temporary fix, bind does not work yet?
+            //    $companyService->saveCompany();
+            //    return $this->redirect()->toRoute('admin_company/default', 
+             //                                     array('action'=>'edit', 
+             //                                           'slugCompanyName' => $companyName), 
+             //                                     array(), false);   
+            //}
+            if ($packetForm->isValid()) {
```

Comment:
You should move this logic to the service. You simply pass the data to the service, which then does all the form checking.